metals/src/main/scala/scala/meta/internal/metals/CodeAction.scala

```diff
@@ -0,0 +1,27 @@
+package scala.meta.internal.metals
+
+import scala.concurrent.{Future, ExecutionContext}
+import org.eclipse.{lsp4j => l}
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.pc.CancelToken
+
+trait CodeAction {
+
+  /**
+   * This should be one of the String constants
+   * listed in [[org.eclipse.lsp4j.CodeActionKind]]
+   */
+  def kind: String
+
+  def contribute(
```

Comment:
Can we reduce the number of parameters in this method? It would be nice to bundle these below into a single class so that we can add more components in the future without updating a ton of signatures

* compilers
* trees
* buffers
* semanticdbs
* symbol search
* definition provider

We should try to design these APIs as if they're public and try to keep them as stable as possible.

Reply:
I thought about this as well, and one alternative way of doing this is to have each CodeAction as a class, which take the necessary “global” parameters upon initialization and the “per request” parameters in the contribute method.

Something like

```
class ImportMissingSymbol(
  compilers: Compilers,
  // ...
) extends CodeAction {
  override def contribute(params: CodeActionParams): Future[List[l.CodeAction]] = // ...
```

Reply:
This would allow each code action to manage their required dependencies independently. I think I was too eager in trying to generalize a common interface.

Reply:
I've done as @gabro suggested. The `CodeAction` trait's `contribute` method now takes only two arguments, the `params: CodeActionParams` and the cancel token. Each code action is now a class that takes its dependencies as constructor parameters.

I agree we could maybe try to group the dependencies into a small number of approachable APIs that abstract over them, but I feel it's too soon to do that. So far we only have 2 code actions, and they each depend on completely different functionality, so it's not yet clear how to define sensible abstractions. Maybe revisit this once we've built a couple more code actions and commonalities start to emerge? @olafurpg